tough/src/editor/targets.rs

```diff
@@ -0,0 +1,536 @@
+        }
+        if let Some(targets) = self.new_targets.as_mut() {
+            targets.remove(name);
+        }
+
+        self
+    }
+
+    /// Remove all targets from this repo
+    pub fn clear_targets(&mut self) -> &mut Self {
+        self.existing_targets
+            .get_or_insert_with(HashMap::new)
+            .clear();
+        self.new_targets.get_or_insert_with(HashMap::new).clear();
+        self
+    }
+
+    /// Set the version
+    pub fn version(&mut self, version: NonZeroU64) -> &mut Self {
+        self.version = Some(version);
+        self
+    }
+
+    /// Set the expiration
+    pub fn expires(&mut self, expires: DateTime<Utc>) -> &mut Self {
+        self.expires = Some(expires);
+        self
+    }
+
+    /// Adds a key to delegations keyids, adds the key to `role` if it is provided
+    pub fn add_key(
+        &mut self,
+        keys: HashMap<Decoded<Hex>, Key>,
+        role: Option<&str>,
+    ) -> Result<&mut Self> {
+        let delegations = self.delegations.as_mut().context(error::NoDelegations)?;
+        let mut keyids = Vec::new();
+        for (keyid, key) in keys {
+            // Check to see if the key is present
+            if delegations
+                .keys
+                .iter()
+                .find(|(_, candidate_key)| key.clone().eq(candidate_key))
+                .is_none()
+            {
+                // Key isn't present yet, so we need to add it
+                delegations.keys.insert(keyid.clone(), key);
+            };
+            keyids.push(keyid.clone());
+        }
+
+        // If a role was provided add keyids to the delegated role
+        if let Some(role) = role {
+            for delegated_role in &mut delegations.roles {
+                if delegated_role.name == role {
+                    delegated_role.keyids.extend(keyids.clone());
+                }
+            }
+            for delegated_role in self.new_roles.get_or_insert(Vec::new()).iter_mut() {
+                if delegated_role.name == role {
+                    delegated_role.keyids.extend(keyids.clone());
+                }
+            }
+        }
+        Ok(self)
+    }
+
+    /// Removes a key from delegations keyids, if a role is specified only removes it from the role
+    pub fn remove_key(&mut self, keyid: &Decoded<Hex>, role: Option<&str>) -> Result<&mut Self> {
+        let delegations = self.delegations.as_mut().context(error::NoDelegations)?;
+        // If a role was provided remove keyid from the delegated role
+        if let Some(role) = role {
+            for delegated_role in &mut delegations.roles {
+                if delegated_role.name == role {
+                    delegated_role.keyids.retain(|key| keyid != key);
+                }
+            }
+        } else {
+            delegations.keys.remove(keyid);
+        }
+        Ok(self)
+    }
+
+    /// Delegates `paths` to `targets` adding a `DelegatedRole` to `new_roles`
+    pub fn delegate_role(
```

Comment:
The comment here is a little confusing. Could we just say `Adds a DelegatedRole `?